File: hub/plantsvc/analytics/growth.py
"""Growth analytics: canopy series, silhouettes, midday droop, RGR + Cohen's d."""

from __future__ import annotations

import json
from typing import Any

import numpy as np
import pandas as pd

from ..config_model import Config
from ..db import Frames
from ..timeutil import iso_utc
from .roster import Roster


def _dawn(fr: Frames) -> pd.DataFrame:
    g = fr.grow
    if g.empty:
        return g
    return g[g.phase == "dawn"] if "phase" in g.columns else g


def _iso(ts) -> str | None:
    return iso_utc(ts.to_pydatetime()) if ts is not None and pd.notna(ts) else None


def canopy_series(fr: Frames, roster: Roster, phase: str = "dawn") -> dict[str, Any]:
    g = fr.grow
    if g.empty:
        return {"pots": []}
    if phase != "all" and "phase" in g.columns:
        g = g[g.phase == phase]
    pots = []
    for p in roster.pots:
        d = g[g.plant_id == p].sort_values("ts")
        if d.empty:
            continue
        pots.append({"plant_id": p, "treat": roster.treat.get(p),
                     "ts": [_iso(t) for t in d.ts],
                     "area_cm2": [None if pd.isna(v) else round(float(v), 2) for v in d.area_cm2]})
    return {"pots": pots}


def silhouettes(fr: Frames, roster: Roster) -> dict[str, Any]:
    dawn = _dawn(fr)
    out: dict[str, Any] = {"pots": [], "not_enough": []}
    if dawn.empty or "contour" not in dawn.columns:
        out["not_enough"] = list(roster.pots)
        return out
    for treat, members in roster.groups.items():
        for p in members:
            d = dawn[dawn.plant_id == p].dropna(subset=["contour"]).sort_values("ts")
            if len(d) < 2:
                out["not_enough"].append(p)
                continue
            new = d.iloc[-1]
            older = d[d.ts <= new.ts - pd.Timedelta(days=1)]
            old = older.iloc[-1] if len(older) else d.iloc[0]
            gap_d = (new.ts - old.ts).total_seconds() / 86400
            gain = 100 * (new.area_px - old.area_px) / old.area_px if old.area_px else None
            base = d.iloc[0]
            span_d = (new.ts - base.ts).total_seconds() / 86400
            tot = (100 * (new.area_px - base.area_px) / base.area_px) if base.area_px else None
            try:
                new_c = json.loads(new.contour)
                old_c = json.loads(old.contour)
            except (TypeError, ValueError):
                out["not_enough"].append(p)
                continue
            lim = float(np.abs(np.array(new_c)).max()) * 1.12 if new_c else 1.0

            def row(r, c):
                return {"ts": _iso(r.ts), "area_cm2": None if pd.isna(r.area_cm2) else round(float(r.area_cm2), 2),
                        "area_px": int(r.area_px), "contour": c}
            out["pots"].append({"plant_id": p, "treat": treat, "new": row(new, new_c),
                                "old": row(old, old_c), "gain_pct": None if gain is None else round(float(gain), 1),
                                "gap_d": round(gap_d, 2),
                                "base": {"ts": _iso(base.ts), "area_cm2": None if pd.isna(base.area_cm2) else round(float(base.area_cm2), 2)},
                                "total_pct": None if tot is None else round(float(tot), 1),
                                "span_d": round(span_d, 2), "lim": round(lim, 1)})
    return out


def droop(fr: Frames, roster: Roster, cfg: Config) -> dict[str, Any]:
    g = fr.grow
    out: dict[str, Any] = {"rows": [], "missing": list(roster.pots), "has_both_phases": False}
    if g.empty or "phase" not in g.columns or not {"dawn", "pm"} <= set(g.phase.unique()):
        return out
    out["has_both_phases"] = True
    g2 = g.copy()
    g2["day"] = g2.ts.dt.tz_convert(cfg.tz).dt.date.astype(str)
    pv = g2.pivot_table(index=["day", "plant_id"], columns="phase", values="area_px").dropna()
    if pv.empty:
        return out
    pv["droop"] = 100 * (pv.dawn - pv.pm) / pv.dawn
    last = pv.reset_index().groupby("plant_id").last().reset_index()
    have = [p for p in roster.pots if p in set(last.plant_id)]
    last = last.set_index("plant_id").loc[have].reset_index() if have else last.iloc[0:0]
    out["rows"] = [{"pot": r.plant_id, "treat": roster.treat.get(r.plant_id), "day": r.day,
                    "dawn_px": int(r.dawn), "pm_px": int(r.pm), "droop_pct": round(float(r.droop), 2)}
                   for r in last.itertuples()]
    out["missing"] = [p for p in roster.pots if p not in have]
    return out


def droop_timeline(fr: Frames, roster: Roster, cfg: Config, days: int = 14) -> dict[str, Any]:
    """Daily droop per pot over the last `days` local days (the '3-b' report item as a series)."""
    g = fr.grow
    if g.empty or "phase" not in g.columns:
        return {"days": [], "pots": []}
    g2 = g.copy()
    g2["day"] = g2.ts.dt.tz_convert(cfg.tz).dt.date.astype(str)
    pv = g2.pivot_table(index=["day", "plant_id"], columns="phase", values="area_px").dropna()
    if pv.empty or "dawn" not in pv.columns or "pm" not in pv.columns:
        return {"days": [], "pots": []}
    pv["droop"] = 100 * (pv.dawn - pv.pm) / pv.dawn
    tbl = pv.droop.unstack("plant_id")
    tbl = tbl.iloc[-days:]
    return {"days": list(tbl.index),
            "pots": [{"plant_id": p, "treat": roster.treat.get(p),
                      "droop_pct": [None if pd.isna(v) else round(float(v), 2) for v in tbl[p]]}
                     for p in roster.pots if p in tbl.columns]}


def rgr(fr: Frames, roster: Roster) -> dict[str, Any]:
    dawn = _dawn(fr)
    rows = []
    for p in roster.pots:
        d = dawn[dawn.plant_id == p].sort_values("ts").dropna(subset=["area_cm2"]) if len(dawn) else dawn
        d = d[d.area_cm2 > 0] if len(d) else d
        if len(d) < 3:
            continue
        t = (d.ts - d.ts.iloc[0]).dt.total_seconds().to_numpy() / 86400
        y = np.log(d.area_cm2.to_numpy(dtype=float))
        slope, intercept = np.polyfit(t, y, 1)
        resid = y - (slope * t + intercept)
        denom = ((t - t.mean()) ** 2).sum()
        se = float(np.sqrt((resid ** 2).sum() / (len(t) - 2) / denom)) if len(t) > 2 and denom > 0 else float("nan")
        ss_tot = ((y - y.mean()) ** 2).sum()
        r2 = float(1 - (resid ** 2).sum() / ss_tot) if ss_tot > 0 else 1.0
        rows.append({"pot": p, "treat": roster.treat.get(p), "rgr": round(float(slope), 5),
                     "se": None if np.isnan(se) else round(se, 5),
                     "ci95": None if np.isnan(se) else [round(float(slope - 1.96 * se), 5), round(float(slope + 1.96 * se), 5)],
                     "r2": round(r2, 4), "n": int(len(t))})
    out: dict[str, Any] = {"pots": rows, "groups": {}, "cohens_d": None, "effect": None,
                           "worst_r2": None, "comparable": False}
    if not rows:
        return out
    r = pd.DataFrame(rows)
    for k in ("stable", "fluct"):
        sub = r[r.treat == k].rgr
        if len(sub):
            out["groups"][k] = {"mean": round(float(sub.mean()), 5),
                                "sd": None if len(sub) < 2 else round(float(sub.std()), 5),
                                "n": int(len(sub))}
    out["worst_r2"] = round(float(r.r2.min()), 4)
    if r.treat.nunique() == 2 and {"stable", "fluct"} <= set(r.treat):
        a, b = r[r.treat == "stable"].rgr, r[r.treat == "fluct"].rgr
        pooled = np.sqrt(((a.std() if len(a) > 1 else 0) ** 2 + (b.std() if len(b) > 1 else 0) ** 2) / 2)
        if pooled and not np.isnan(pooled):
            d_eff = float((a.mean() - b.mean()) / pooled)
            out["cohens_d"] = round(d_eff, 3)
            out["effect"] = "large" if abs(d_eff) >= .8 else "medium" if abs(d_eff) >= .5 else "small"
        out["comparable"] = True
    return out

File: web/src/mock/growth.ts
/* TS ports of analytics/growth.py and analytics/irrigation.py. */
import type { Canopy, Droop, DroopTimeline, PlantConfig, PumpRecentRow, Rgr, SilFrame, Silhouettes, Water } from '@/api/types'
import { mean, sd } from './analytics'
import type { Frames, GrowRow, Roster } from './frames'
import { iso, localDay, round } from './time'

const DAY = 86_400_000
const byTs = <T extends { ts: number }>(a: T, b: T) => a.ts - b.ts
const dawnRows = (fr: Frames) => fr.grow.filter((r) => r.phase === 'dawn')
const potRows = (rows: GrowRow[], p: string) => rows.filter((r) => r.plant_id === p).sort(byTs)

export function canopySeries(fr: Frames, roster: Roster, phase = 'dawn'): Pick<Canopy, 'pots'> {
  const g = phase === 'all' ? fr.grow : fr.grow.filter((r) => r.phase === phase)
  const pots: Canopy['pots'] = []
  for (const p of roster.pots) {
    const d = potRows(g, p)
    if (!d.length) continue
    pots.push({ plant_id: p, treat: roster.treat[p] ?? null, ts: d.map((r) => iso(r.ts)), area_cm2: d.map((r) => (r.area_cm2 === null ? null : round(r.area_cm2, 2))) })
  }
  return { pots }
}

export function silhouettes(fr: Frames, roster: Roster): Pick<Silhouettes, 'pots' | 'not_enough'> {
  const dawn = dawnRows(fr)
  const out: Pick<Silhouettes, 'pots' | 'not_enough'> = { pots: [], not_enough: [] }
  if (!dawn.length) { out.not_enough = [...roster.pots]; return out }
  const frame = (r: GrowRow): SilFrame => ({ ts: iso(r.ts), area_cm2: r.area_cm2 === null ? null : round(r.area_cm2, 2), area_px: r.area_px, contour: r.contour ?? [] })
  for (const [treat, members] of Object.entries(roster.groups)) {
    for (const p of members) {
      const d = potRows(dawn, p).filter((r) => r.contour !== null)
      if (d.length < 2) { out.not_enough.push(p); continue }
      const nw = d[d.length - 1]
      const older = d.filter((r) => r.ts <= nw.ts - DAY)
      const old = older.length ? older[older.length - 1] : d[0]
      const base = d[0]
      const lim = nw.contour?.length ? Math.max(...nw.contour.flat().map(Math.abs)) * 1.12 : 1
      out.pots.push({
        plant_id: p, treat, new: frame(nw), old: frame(old),
        gain_pct: old.area_px ? round((100 * (nw.area_px - old.area_px)) / old.area_px, 1) : null, gap_d: round((nw.ts - old.ts) / DAY, 2),
        base: { ts: iso(base.ts), area_cm2: base.area_cm2 === null ? null : round(base.area_cm2, 2) },
        total_pct: base.area_px ? round((100 * (nw.area_px - base.area_px)) / base.area_px, 1) : null, span_d: round((nw.ts - base.ts) / DAY, 2), lim: round(lim, 1),
      })
    }
  }
  return out
}

/** pivot_table(index=[day, plant_id], columns=phase, values=area_px).dropna() -> day -> pot -> droop % */
function droopPivot(fr: Frames, tz: string): Map<string, Map<string, { dawn: number; pm: number; droop: number }>> {
  const acc = new Map<string, Map<string, { dawn: number[]; pm: number[] }>>()
  for (const r of fr.grow) {
    const day = localDay(r.ts, tz)
    let byPot = acc.get(day)
    if (!byPot) { byPot = new Map(); acc.set(day, byPot) }
    let cell = byPot.get(r.plant_id)
    if (!cell) { cell = { dawn: [], pm: [] }; byPot.set(r.plant_id, cell) }
    cell[r.phase].push(r.area_px)
  }
  const out = new Map<string, Map<string, { dawn: number; pm: number; droop: number }>>()
  for (const day of [...acc.keys()].sort()) {
    const m = new Map<string, { dawn: number; pm: number; droop: number }>()
    for (const [pid, c] of acc.get(day)!) {
      if (!c.dawn.length || !c.pm.length) continue
      const dawn = mean(c.dawn), pm = mean(c.pm)
      m.set(pid, { dawn, pm, droop: (100 * (dawn - pm)) / dawn })
    }
    if (m.size) out.set(day, m)
  }
  return out
}

export function droop(fr: Frames, roster: Roster, cfg: PlantConfig): Pick<Droop, 'rows' | 'missing' | 'has_both_phases'> {
  const out: Pick<Droop, 'rows' | 'missing' | 'has_both_phases'> = { rows: [], missing: [...roster.pots], has_both_phases: false }
  const phases = new Set(fr.grow.map((r) => r.phase))
  if (!fr.grow.length || !phases.has('dawn') || !phases.has('pm')) return out
  out.has_both_phases = true
  const pv = droopPivot(fr, cfg.tz)
  const lastDay = new Map<string, string>()
  for (const [day, m] of pv) for (const pid of m.keys()) lastDay.set(pid, day)
  const have = roster.pots.filter((p) => lastDay.has(p))
  out.rows = have.map((p) => {
    const day = lastDay.get(p)!, c = pv.get(day)!.get(p)!
    return { pot: p, treat: roster.treat[p] ?? null, day, dawn_px: Math.floor(c.dawn), pm_px: Math.floor(c.pm), droop_pct: round(c.droop, 2) }
  })
  out.missing = roster.pots.filter((p) => !lastDay.has(p))
  return out
}

export function droopTimeline(fr: Frames, roster: Roster, cfg: PlantConfig, days = 14): Pick<DroopTimeline, 'days' | 'pots'> {
  if (!fr.grow.length) return { days: [], pots: [] }
  const pv = droopPivot(fr, cfg.tz)
  const allDays = [...pv.keys()].slice(-days)
  if (!allDays.length) return { days: [], pots: [] }
  const seen = new Set<string>()
  for (const m of pv.values()) for (const p of m.keys()) seen.add(p)
  return {
    days: allDays,
    pots: roster.pots.filter((p) => seen.has(p)).map((p) => ({ plant_id: p, treat: roster.treat[p] ?? null, droop_pct: allDays.map((d) => { const v = pv.get(d)!.get(p); return v === undefined ? null : round(v.droop, 2) }) })),
  }
}

export function rgr(fr: Frames, roster: Roster): Pick<Rgr, 'pots' | 'groups' | 'cohens_d' | 'effect' | 'worst_r2' | 'comparable'> {
  const dawn = dawnRows(fr)
  const rows: Rgr['pots'] = []
  for (const p of roster.pots) {
    const d = potRows(dawn, p).filter((r) => r.area_cm2 !== null && r.area_cm2 > 0)
    if (d.length < 3) continue
    const t = d.map((r) => (r.ts - d[0].ts) / DAY), y = d.map((r) => Math.log(r.area_cm2!))
    const tm = mean(t), ym = mean(y)
    const sxx = t.reduce((a, x) => a + (x - tm) ** 2, 0)
    const slope = t.reduce((a, x, i) => a + (x - tm) * (y[i] - ym), 0) / sxx
    const icpt = ym - slope * tm
    const sse = y.reduce((a, v, i) => a + (v - (slope * t[i] + icpt)) ** 2, 0)
    const se = t.length > 2 && sxx > 0 ? Math.sqrt(sse / (t.length - 2) / sxx) : NaN
    const sst = y.reduce((a, v) => a + (v - ym) ** 2, 0)
    const r2 = sst > 0 ? 1 - sse / sst : 1
    rows.push({ pot: p, treat: roster.treat[p] ?? null, rgr: round(slope, 5), se: Number.isNaN(se) ? null : round(se, 5), ci95: Number.isNaN(se) ? null : [round(slope - 1.96 * se, 5), round(slope + 1.96 * se, 5)], r2: round(r2, 4), n: t.length })
  }
  const out: ReturnType<typeof rgr> = { pots: rows, groups: {}, cohens_d: null, effect: null, worst_r2: null, comparable: false }
  if (!rows.length) return out
  for (const k of ['stable', 'fluct']) {
    const sub = rows.filter((r) => r.treat === k).map((r) => r.rgr)
    if (sub.length) out.groups[k] = { mean: round(mean(sub), 5), sd: sub.length < 2 ? null : round(sd(sub), 5), n: sub.length }
  }
  out.worst_r2 = round(Math.min(...rows.map((r) => r.r2)), 4)
  if (out.groups.stable && out.groups.fluct) {
    const a = rows.filter((r) => r.treat === 'stable').map((r) => r.rgr), b = rows.filter((r) => r.treat === 'fluct').map((r) => r.rgr)
    const pooled = Math.sqrt(((a.length > 1 ? sd(a) : 0) ** 2 + (b.length > 1 ? sd(b) : 0) ** 2) / 2)
    if (pooled && !Number.isNaN(pooled)) {
      const dEff = (mean(a) - mean(b)) / pooled
      out.cohens_d = round(dEff, 3)
      out.effect = Math.abs(dEff) >= 0.8 ? 'large' : Math.abs(dEff) >= 0.5 ? 'medium' : 'small'
    }
    out.comparable = true
  }
  return out
}

// ---- irrigation.py --------------------------------------------------------------
export const ML_PER_SEC = 10

export function pumpRecent(fr: Frames, roster: Roster, n = 5): { rows: PumpRecentRow[] } {
  const rows = [...fr.pump].sort((a, b) => b.ts - a.ts).slice(0, n).map((r) => {
    const before = r.soil_before === null ? null : round(r.soil_before, 1), after = r.soil_after === null ? null : round(r.soil_after, 1)
    return {
      ts: iso(r.ts), ago_h: fr.now === null ? null : Math.floor((fr.now - r.ts) / 3_600_000), pot: r.plant_id.toUpperCase(), plant_id: r.plant_id,
      treat: roster.treat[r.plant_id]?.toUpperCase() ?? null, pump_s: r.dur_ms === null ? null : round(r.dur_ms / 1000, 1),
      before, after, rise: before === null || after === null ? null : round(after - before, 1), reason: r.reason === null ? null : r.reason.replace(/_/g, ' '),
    }
  })
  return { rows }
}

export function waterSummary(fr: Frames, roster: Roster, tz: string, days = 14): Pick<Water, 'ml_per_s' | 'groups' | 'pots' | 'daily'> {
  const out: Pick<Water, 'ml_per_s' | 'groups' | 'pots' | 'daily'> = { ml_per_s: ML_PER_SEC, groups: {}, pots: [], daily: { days: [], groups: {} } }
  const p = fr.pump.filter((r) => r.dur_ms !== null && roster.treat[r.plant_id] !== undefined).map((r) => ({ ...r, ml: (r.dur_ms! / 1000) * ML_PER_SEC, treat: roster.treat[r.plant_id] }))
  if (!p.length) return out
  for (const k of [...new Set(p.map((r) => r.treat))].sort()) {
    const sub = p.filter((r) => r.treat === k), tot = sub.reduce((a, r) => a + r.ml, 0), np = new Set(sub.map((r) => r.plant_id)).size
    out.groups[k] = { total_ml: round(tot, 1), events: sub.length, pots: np, ml_per_pot: round(tot / Math.max(1, np), 1) }
  }
  for (const pid of roster.pots) {
    const sub = p.filter((r) => r.plant_id === pid).sort(byTs)
    if (!sub.length) continue
    const gaps = sub.slice(1).map((r, i) => (r.ts - sub[i].ts) / DAY)
    out.pots.push({ plant_id: pid, treat: roster.treat[pid] ?? null, events: sub.length, total_ml: round(sub.reduce((a, r) => a + r.ml, 0), 1), interval_d: gaps.map((g) => round(g, 2)), mean_interval_d: gaps.length ? round(mean(gaps), 2) : null, last_interval_d: gaps.length ? round(gaps[gaps.length - 1], 2) : null })
  }
  const daily = new Map<string, Record<string, number>>()
  for (const r of p) { const d = localDay(r.ts, tz); const row = daily.get(d) ?? {}; row[r.treat] = (row[r.treat] ?? 0) + r.ml; daily.set(d, row) }
  const ds = [...daily.keys()].sort().slice(-days)
  out.daily = { days: ds, groups: Object.fromEntries(Object.keys(out.groups).map((k) => [k, ds.map((d) => round(daily.get(d)![k] ?? 0, 1))])) }
  return out
}
